### PR TITLE
hub/agent: consolidate upgrade implementations

### PR DESCRIPTION
**Pull request by jchampio**
The first commit adds a new upgrade API for both the hub and agent to use, which looks a little like this in practice:
```go
segmentPair := upgrade.SegmentPair{
    Source: &upgrade.Segment{sourceBinDir, segment.OldDataDir, dbid, int(segment.OldPort)},
    Target: &upgrade.Segment{targetBinDir, segment.NewDataDir, dbid, int(segment.NewPort)},
}

err := upgrade.Run(segmentPair,
    upgrade.WithExecCommand(execCommand),
    upgrade.WithWorkDir(segment.WorkDir),
    upgrade.WithSegmentMode(),
)
```

The second commit makes use of this API throughout the code base and removes redundant tests. One business logic change is that master upgrade logs are now correctly `--retain`ed due to the consolidation of code.

### CODE REVIEW

File: upgrade/run.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	return cmd.Run()
+}
+
+// Option configures the way Run executes pg_upgrade.
+type Option func(*optionList)
+
+// WithWorkDir configures the working directory for pg_upgrade. Log files will
+// be retained in this directory. If this option is unset or empty, the calling
+// process's working directory will be used.
+func WithWorkDir(wd string) Option {
```
Comment (berlin-ab) [inline upgrade/run.go]: Cool pattern with these builder-style options. Do we get any benefit from the delayed execution? Why not build up these options as state on a struct?

File: upgrade/run.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	}
+}
+
+// optionList holds the combined result of all possible Options. Zero values
+// represent the default settings.
+type optionList struct {
+	Dir            string
+	CheckOnly      bool
+	ExecCommand    func(string, ...string) *exec.Cmd
+	ExecCommandSet bool // was ExecCommand explicitly set?
```
Comment (kalensk) [inline upgrade/run.go]: Maybe I need to stare at the code longer, but why is this needed? Shouldn't simply setting the ExecCommand be enough? Also, why is ExecCommand even needed on the optionList, rather than simply doing something like this directly:
```
var execCommand = exec.Command

cmd := execCommand(path, args...)
```

File: upgrade/run.go (hunk 1 above)
Comment (jchampio) [reply upgrade/run.go]: I started to write up an explanation and then realized that [Dave Cheney summarizes it the best in the place where I got the idea](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

File: upgrade/run.go (hunk 2 above)
Comment (jchampio) [reply upgrade/run.go]: Good questions, and maybe we need to add some explanation in the code. I'll answer the second one first.
> Also, why is ExecCommand even needed on the optionList, rather than simply doing something like this directly: [...]

Problem with that is that you can no longer stub out the dependency across packages. So if you're running Hub tests and one of the functions you're testing needs to call upgrade, you can't access `upgrade.execCommand` to be able to control it (i.e. prevent upgrade from actually running). So we borrowed the gRPC idea of allowing an option to inject a dependency, similar to `WithContextDialer()`.

> Shouldn't simply setting the ExecCommand be enough?

That was what I thought too. Then @bhuvnesh2703 asked "so what happens if you don't set up a mock command in the tests?" and I answered "it'll crash to let us know there's a problem, because we set it to `nil` at the beginning of the tests, see?"

And then of course it did _not_ crash, and instead accidentally executed `pg_upgrade`, because we took `nil` to mean "use the default". Having a separate boolean allows us to explicitly `nil` out the `MockCommand` implementation.

Comment (kalensk) [reply upgrade/run.go]: Thanks for the explanation! 

Talking out loud here...It would be nice if the `CheckOnly` bool was not needed. That is, I was hoping for defaulting to null. And in the production code use the correct execCommand and for tests use the fake execCommand, and if its not set it will fail since the default is null. However, if one was to implement that there would probably be no way for the tests to "fail" and not use the real pg_upgrade if one forgets to set the fake pg_upgrade execCommand when writing a test. Unfortunate.

File: upgrade/run_test.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+	t.Run("unsets PGPORT and PGHOST", func(t *testing.T) {
+		// Set our environment.
+		os.Setenv("PGPORT", "5432")
+		os.Setenv("PGHOST", "localhost")
+		defer func() {
+			os.Unsetenv("PGPORT")
+			os.Unsetenv("PGHOST")
+		}()
+
```
Comment (dkrieger-pvtl) [inline upgrade/run_test.go]: It's a pedantic point, but you risk polluting other tests in this package by changing their `PGPORT` and `PGHOST` values to the unset state.  As far as I can quickly tell, `go test` tests all run in the same process, hence my concern.  Instead, use this idiom:
```
	oldVal, isSet := os.LookupEnv("VAR")
	defer func() {
		if isSet {
			os.Setenv("VAR", oldVal)
		}
	}()
	err = os.Setenv("VAR", "5432")
```

File: agent/upgrade_primaries.go [hunk 4]
```diff
@@ -39,14 +37,8 @@ func UpgradePrimary(sourceBinDir string, targetBinDir string, dataDirPairs []*id
 	segments := make([]Segment, 0, len(dataDirPairs))
 
 	for _, dataPair := range dataDirPairs {
-		err := utils.System.MkdirAll(dataPair.NewDataDir, 0700)
```
Comment (dkrieger-pvtl) [inline agent/upgrade_primaries.go]: Where is the target segment data directory created now?  Or was it always unnecessary?

**Comment by dkrieger-pvtl** [comment]
This needs to be rebased with origin/master(see cli/commanders/initialize.go).

File: upgrade/run.go (hunk 1 above)
Comment (dkrieger-pvtl) [reply upgrade/run.go]: This is a pretty cool pattern.  And, unlike the blog post, you don't expose internal implementation details but set an Options struct.

File: upgrade/run.go (hunk 2 above)
Comment (dkrieger-pvtl) [reply upgrade/run.go]: @kalensk why would you not need `checkOnly` as an option?  Sometimes we run

File: upgrade/run_test.go (hunk 3 above)
Comment (bhuvnesh2703) [reply upgrade/run_test.go]: applied feedback.

**Comment by bhuvnesh2703** [comment]
Rebased and applied the feedback required by @dkrieger-pvtl
